### PR TITLE
FOGL-8753 : Implemented user blocking for incorrect password

### PR DESCRIPTION
**Pull request by gnandan**

### CODE REVIEW

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:687]: Why do we need payload? As endpoint says unblock! It should be without payload.

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:697]: Can we block Super Admin user? I believe NO

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:704]: We have already a import datetime in file, use from it

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:715]: Add a note: Why failed_attempt reset to 0?

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:711]: If 'rows' Key is not available then? raise Storage Server Exception

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:721]: blocked or unblocked?

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:725]: What to enable?

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:350]: Space between operator

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:380]: Use DATE_FORMAT variable here

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:395]: Declare MAX_ATTEMPTS above

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:410]: Raise user alert as well

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:394]: admin user?

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:404]: terminate the sentence

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:414]: terminate the sentence

File: scripts/plugins/storage/postgres/init.sql
Comment (ashish-jabble) [inline scripts/plugins/storage/postgres/init.sql:606]: with timezone?

**Review by ashish-jabble — CHANGES_REQUESTED** [review]
Dry run review!
Unit tests are failing and good to have new tests for existing API as user update have been touched!

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:384]: diff variable unused?

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:350]: It's default value always from user; You can remove these from here

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:395]: audit instantiate should move just before it's usage

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:615]: Why this change?

File: scripts/plugins/storage/postgres/init.sql
Comment (MarkRiddoch) [reply scripts/plugins/storage/postgres/init.sql:606]: Probably best tp use etc as the user might change timezones, especially should not be based on the timezone of the caller of the API.

**Review by MarkRiddoch — COMMENTED** [review]
Looks pretty good - we should probably address the timezone issue however.

File: scripts/plugins/storage/postgres/init.sql
Comment (gnandan) [reply scripts/plugins/storage/postgres/init.sql:606]: @MarkRiddoch While login user can only provide id and password. There is no way user can provide any information related to time. API does all the calculation for block time on the basis of information stored into the database. So we are safe here even if user changes timezone. so is it good now?

**Review by MarkRiddoch — DISMISSED** [review]

**Comment by ashish-jabble** [comment]
Audit API tests are also failing with the given change

**Comment by ashish-jabble** [comment]
VERSION file is also missing. It should be with version 72

**Comment by praveen-garg** [comment]
@mshariq-nerd Please test and review per GUI needs so that we can expedite merge.

**Comment by mshariq-nerd** [comment]
@gnandan, As per Jira
> Add an alert whenever an account is blocked for 24 hours.

This seems to be missing. Alert Manager shows this?

**Comment by mshariq-nerd** [comment]
@gnandan 
What will happen of a active session if logged in with same credentials in different browser?
For example A user already logged in a browser X  and try to login into the same account from the different browser Y, multiple times Or someone login impersonating me from Y browser now user will be blocked at X browser?
cc/ @MarkRiddoch

File: scripts/plugins/storage/sqlite/upgrade/72.sql
Comment (ashish-jabble) [inline scripts/plugins/storage/sqlite/upgrade/72.sql:2]:
```
ALTER TABLE fledge.users ADD COLUMN block_until DATETIME DEFAULT NULL;

UPDATE fledge.users SET block_until=current_timestamp;
```

Comment (praveen-garg) [reply scripts/plugins/storage/sqlite/upgrade/72.sql:2]: Blocked Until should be default as null.

**Comment by MarkRiddoch** [comment]
> @gnandan What will happen of a active session if logged in with same credentials in different browser? For example A user already logged in a browser X and try to login into the same account from the different browser Y, multiple times Or someone login impersonating me from Y browser now user will be blocked at X browser? cc/ @MarkRiddoch

Any existing login by the user should not be affected by the blocking the user.

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:335]: API should always return in UTC timestamp.

File: scripts/plugins/storage/sqlite/upgrade/72.sql
Comment (ashish-jabble) [inline scripts/plugins/storage/sqlite/upgrade/72.sql:3]: Sqlite: CURRENT_TIMESTAMP is in GMT, not the timezone of the machine.

`UPDATE users SET block_until=(SELECT STRFTIME('%Y-%m-%d %H:%M:%f', 'NOW', 'localtime'))`

**Review by MarkRiddoch — COMMENTED** [review]
Please resolve timezone issue and then I think this is probably ok

File: scripts/plugins/storage/sqlite/upgrade/72.sql
Comment (praveen-garg) [reply scripts/plugins/storage/sqlite/upgrade/72.sql:3]: Why are we not setting null? TS should be only if blocked.

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:711]: Return value is expected here and we should move further only if success.

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:737]: We do not need to fetch all of the columns here; Simple with one column and check if storage returns any row then move else raise the condition

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:745]: Do we really want to reset block_until column with current datetime? Or with default NULL?

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:335]: For user profile we do not need `status and blockUntil` KV pair's, this can be removed from here.

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:357]: As these both keys are dependent and only exists if block_until. So we may add either with
```
"block": {"status": true/false, "until": "YYY-MM-DD...."}

Or

only with single key "blockUntil"
```

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:196]: `unblock_user` key in user_data in what scenario?

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:383]: Seems to be an unused variable

File: python/fledge/services/core/api/auth.py
Comment (MarkRiddoch) [reply python/fledge/services/core/api/auth.py]: No we do not currently. We should make this any user with admin right in future and allowing blocking of admin users as long as there is always one unblocked admin user. This is not for the current change however.

**Review by MarkRiddoch — APPROVED** [review]
Looks fine, ideally address comments from Ashish before checkin

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:737]: This will break in PostgreSQL.

Comment (MarkRiddoch) [reply python/fledge/services/core/api/auth.py:737]: > This will break in PostgreSQL.

We must fix this

Comment (ashish-jabble) [reply python/fledge/services/core/api/auth.py:737]: To insert null value to the database from Python, just use  'None' as it corresponds to NULL type.

File: python/fledge/services/core/api/auth.py
Comment (ashish-jabble) [inline python/fledge/services/core/api/auth.py:699]: type cast of user_id only once

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:381]:
```
block_until_datetime_formatter = datetime.strptime(block_until, DATE_FORMAT)
current_datetime_formatter = datetime.strptime(curr_time, DATE_FORMAT)
```

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:387]: `hours_left = "{} hour ".format(hours) if hours == 1 else "{} hours ".format(hours)`

File: python/fledge/services/core/user_model.py
Comment (ashish-jabble) [inline python/fledge/services/core/user_model.py:392]: This can also be break into single statement

**Review by ashish-jabble — APPROVED** [review]
Code review done!

**Review by MarkRiddoch — APPROVED** [review]

**Pull request merged**
